src/tree.ml

```diff
@@ -0,0 +1,711 @@
+      let ppit oc (a, b) =
+        Format.fprintf oc "@[(%a@@%d)@]" pp_tvs (Array.to_list b) a
+      in
+      let pp_sep oc () = Format.pp_print_string oc "::" in
+      Format.fprintf oc "@[<v>@[%a@]@]"
+        (Format.pp_print_list ~pp_sep ppit) (IntMap.bindings fv_conds)
+    in
+    let pp_partials oc partials =
+      let pp_sep oc _ = Format.pp_print_string oc ";" in
+      let pp_part oc (i, j) = Format.fprintf oc "@[(%d↦%d)@]" i j in
+      Format.fprintf oc "@[<h>%a@]" (Format.pp_print_list ~pp_sep pp_part)
+        (IntMap.bindings partials)
+    in
+    let pp_available oc available =
+      let pp_av oc (i, j) = Format.fprintf oc "@[%d≡%d@]" i j in
+      let pp_sep = Format.pp_print_cut in
+      Format.fprintf oc "@[<hov>%a@]" (Format.pp_print_list ~pp_sep pp_av)
+        (PSet.elements available)
+    in
+    Format.fprintf oc "@[<hov>%a/@,%a/@,%a@]" pp_fv pool.fv_conds
+      pp_partials pool.variables pp_available pool.nl_conds
+
+  (** [is_empty pool] tells whether the pool of constraints is empty. *)
+  let is_empty : t -> bool = fun pool ->
+    PSet.is_empty pool.nl_conds && IntMap.is_empty pool.fv_conds
+
+  (** [register_nl slot i pool] registers the fact that the slot [slot] in the
+      [vars] array correspond to a term stored at index [i] in the environment
+      for the RHS. The first time that such a slot is associated to [i], it is
+      registered to serve as a reference point for testing convertibility when
+      (and if) another such slot is (ever) encountered. When that is the case,
+      a convertibility constraint is registered between the term stored in the
+      slot [slot] and the term stored in the reference slot. *)
+  let register_nl : int -> int -> t -> t = fun slot i pool ->
+    try
+      (* We obtain the reference slot (if any). *)
+      let r_slot = IntMap.find i pool.variables in
+      let cond = if r_slot < slot then (r_slot, slot) else (slot, r_slot) in
+      { pool with nl_conds = PSet.add cond pool.nl_conds }
+    with Not_found ->
+      (* First occurence of [i], register the slot as a point of reference. *)
+      { pool with variables = IntMap.add i slot pool.variables }
+
+  (** [register_fv slot xs pool] registers a free variables constraint for the
+      variables in [xs] on the slot [slot] of the [vars] array in [pool]. *)
+  let register_fv : int -> tvar array -> t -> t = fun i vs pool ->
+    { pool with fv_conds = IntMap.add i vs pool.fv_conds }
+
+  (** [constrained_nl i pool] tells whether index [i] in the RHS's environment
+      has already been associated to a variable of the [vars] array. *)
+  let constrained_nl : int -> t -> bool = fun slot pool ->
+    IntMap.mem slot pool.variables
+
+  (** [is_contained cond pool] tells whether the condition [cond] is contained
+      in the pool [pool]. *)
+  let is_contained : tree_cond -> t -> bool = fun cond pool ->
+    match cond with
+    | CondNL(i,j) -> PSet.mem (if i < j then (i,j) else (j,i)) pool.nl_conds
```

Comment:
When I talked about normalization in my previous comment, I'm a bit concerned here. For instance, we might have a convertibility constraint between `i` and `j`, but we might not have `(i,j)` nor `(j,i)` in the set of constraints. We may only have `(k,i)` and `(k,j)` for some slot `k`.

Reply:
I don't think we can have this situation (`(k, i), (k, j), (i, j)`) since the reference slot is given by the pattern in the `Patt(Some(s), _, _)`.  Conditions involving slots `i`, `k` and `j` will be added with `s` as argument (which is unique, otherwise there is no non linearity constraint).  First pattern added will create the binding `s -> i` in `pool.variables` and other will retrieve the correct slot (here `i`) using that mapping.  Is there something I miss?